### PR TITLE
Added JSON.ARRPOP command

### LINKED ISSUE #489: Add support for command `JSON.ARRPOP`
Add support for the `JSON.ARRPOP` command in DiceDB similar to the [`JSON.ARRPOP` command in Redis](https://redis.io/docs/latest/commands/command-list/). Please refer to the following commit in Redis to understand the implementation specifics - [source](https://github.com/redis/redis/tree/f60370ce28b946c1146dcea77c9c399d39601aaa).

Write unit and integration tests for the command referring to the tests written in the [Redis codebase 7.2.5](https://github.com/redis/redis/tree/f60370ce28b946c1146dcea77c9c399d39601aaa). For integration tests, you can refer to the `tests` folder. Note: they have used `TCL` for the test suite, and we need to port that to our way of writing integration tests using the relevant helper methods. Please refer to our [tests directory](https://github.com/DiceDB/dice/tree/master/tests).

For the command, benchmark the code and measure the time taken and memory allocs using [`benchmem`](https://blog.logrocket.com/benchmarking-golang-improve-function-performance/) and try to keep them to the bare minimum.

### PR DESCRIPTION
**Pull request by psrvere**
Closes #489 

Implemented JSON.ARRPOP command.

### CODE REVIEW

File: internal/eval/eval.go [hunk 1]
```diff
@@ -587,6 +587,118 @@ func evalJSONARRLEN(args []string, store *dstore.Store) []byte {
...
+
+		// if index is negative, change it to equivalent positive index
+		if idx < 0 {
+			// if index is out of bound then limit it to the first index
+			if idx < -len(arr) {
+				idx = 0
+			} else {
+				idx = len(arr) + idx
+			}
+		}
```
Comment (lucifercr07) [inline internal/eval/eval.go:689]: Shall we move this to a separate function call?

File: internal/eval/eval.go [hunk 2]
```diff
@@ -587,6 +587,118 @@ func evalJSONARRLEN(args []string, store *dstore.Store) []byte {
...
+		arr, ok := result.([]any)
+		// if value can not be converted to array, it is of another type
+		// return nill in this case similar to redis
+		// also, return nil if array is empty
+		if !ok || len(arr) == 0 {
+			popElem := clientio.RespNIL
+			popArr = append(popArr, popElem)
+			continue
+		}
+		popElem := popElementAndUpdateArray(arr, index, key, store)
```
Comment (lucifercr07) [inline internal/eval/eval.go:654]: Shall we handle `nil` return here?

**Review by lucifercr07 — CHANGES_REQUESTED** [review]

File: internal/eval/eval.go [hunk 3]
```diff
@@ -587,6 +587,118 @@ func evalJSONARRLEN(args []string, store *dstore.Store) []byte {
...
+				idx = len(arr) + idx
+			}
+		}
+	}
+
+	popElem := arr[idx]
+	arr = append(arr[:idx], arr[idx+1:]...)
+
+	// save the remaining array
+	newObj := store.NewObj(arr, -1, object.ObjTypeJSON, object.ObjEncodingJSON)
```
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/eval/eval.go]: Yes, you are right. Also, I had to return updated array with `popElementAndUpdateArray` to get it right.

File: internal/eval/eval.go (hunk 1 above)
Comment (psrvere) [reply internal/eval/eval.go]: Done.

File: internal/eval/eval.go (hunk 2 above)
Comment (psrvere) [reply internal/eval/eval.go]: added an error in return and handled that.

**Comment by psrvere** [comment]
Integration tests failing, looking into it

**Comment by psrvere** [comment]
@lucifercr07 @JyotinderSingh - fixed everything. Please review.

File: internal/eval/eval.go [hunk 4]
```diff
@@ -588,6 +588,142 @@ func evalJSONARRLEN(args []string, store *dstore.Store) []byte {
...
+func evalJSONARRPOP(args []string, store *dstore.Store) []byte {
+	if len(args) < 1 {
+		return diceerrors.NewErrArity("json.arrpop")
+	}
+	key := args[0]
+
+	var path string = defaultRootPath
+	if len(args) >= 2 {
+		path = args[1]
+	}
```
Comment (jujiale) [inline internal/eval/eval.go:600]: this pr looks good. I have a few comments.  
in redis. if the key neither is an array nor exist . should raise an error.
`

    127.0.0.1:6379> JSON.SET user $ '{"a":[{"b":2},{"g":[1,2]},3]}'
    OK
    127.0.0.1:6379> JSON.ARRPOP user1
    (error) ERR Path '$' does not exist or not an array
    127.0.0.1:6379> JSON.ARRPOP user
    (error) ERR Path '$' does not exist or not an array
    127.0.0.1:6379> JSON.SET user $ '[{"b":2},{"g":[1,2]},3]'
    OK
    127.0.0.1:6379> JSON.ARRPOP user
    "3"
  

`

**Review by lucifercr07 — COMMENTED** [review]
@psrvere LGTM, please have a look to comments from @jujiale once. Thanks for contributing.

File: internal/eval/eval.go [hunk 5]
```diff
@@ -588,6 +588,142 @@ func evalJSONARRLEN(args []string, store *dstore.Store) []byte {
...
+	jsonData := obj.Value
+	_, err := sonic.Marshal(jsonData)
+	if err != nil {
+		return diceerrors.NewErrWithMessage("Existing key has wrong Dice type")
+	}
+
+	if path == defaultRootPath {
+		arr, ok := jsonData.([]any)
+		// if value can not be converted to array, it is of another type
+		// return nill in this case similar to redis
```
[the remark this answers is not in the document]
Comment (psrvere) [reply internal/eval/eval.go]: fixed.

File: internal/eval/eval.go (hunk 4 above)
Comment (psrvere) [reply internal/eval/eval.go:600]: Thanks @jujiale. Fixed this.

**Comment by psrvere** [comment]
@lucifercr07 - thanks! addressed jujiale's comments.

**Comment by lucifercr07** [comment]
LGTM, merging.
cc: @JyotinderSingh

**Pull request merged**